uchiwa/daemon/daemon.go

```diff
@@ -66,124 +97,247 @@ func (d *Daemon) buildData() {
+	dc.Metrics["stashes"] = len(d.snapshot.Stashes)
+
+	// update datacenter in the Daemon scope
+	f.mutex.Lock()
+	f.data.Dc = append(f.data.Dc, dc)
+	f.data.Health.Sensu[f.datacenter.Name] = d.determineHealth()
+	f.data.Stashes = append(f.data.Stashes, d.snapshot.Stashes...)
+	f.data.Silenced = append(f.data.Silenced, d.snapshot.Silenced...)
+	f.data.Checks = append(f.data.Checks, d.snapshot.Checks...)
+	f.data.Clients = append(f.data.Clients, d.snapshot.Clients...)
+	f.data.Events = append(f.data.Events, d.snapshot.Events...)
+	f.data.Aggregates = append(f.data.Aggregates, d.snapshot.Aggregates...)
+
+	if f.enterprise {
+		f.data.SERawMetrics.Clients = append(f.data.SERawMetrics.Clients, d.metrics.Clients...)
+		f.data.SERawMetrics.Events = append(f.data.SERawMetrics.Events, d.metrics.Events...)
+		f.data.SERawMetrics.KeepalivesAVG60 = append(f.data.SERawMetrics.KeepalivesAVG60, d.metrics.KeepalivesAVG60...)
+		f.data.SERawMetrics.Requests = append(f.data.SERawMetrics.Requests, d.metrics.Requests...)
+		f.data.SERawMetrics.Results = append(f.data.SERawMetrics.Results, d.metrics.Results...)
+	}
+
+	f.mutex.Unlock()
+
+	logger.Infof("Updated the datacenter %s", f.datacenter.Name)
+
+	f.wg.Done()
+}
+
+func (d *DatacenterSnapshotFetcher) determineHealth() structs.SensuHealth {
+	if !d.snapshot.Info.Redis.Connected {
+		return structs.SensuHealth{Output: "Not connected to Redis", Status: 1}
+	} else if !d.snapshot.Info.Transport.Connected {
+		return structs.SensuHealth{Output: "Not connected to the transport", Status: 1}
+	} else if d.snapshot.Error != "" {
+		return structs.SensuHealth{Output: d.snapshot.Error, Status: 2}
+	}
+
+	return structs.SensuHealth{Output: "ok", Status: 0}
+}
+
+func (d *DatacenterSnapshotFetcher) fetchStashes() {
+	stashes, err := d.datacenter.GetStashes()
+	d.mutex.Lock()
+	if err != nil {
+		logger.Debug(err)
+		logger.Warningf("Connection failed to the datacenter %s", d.datacenter.Name)
+		d.snapshot.Error = err.Error()
+	}
 
-		// build datacenter
-		dc := d.buildDatacenter(&datacenter.Name, info)
-		dc.Metrics["aggregates"] = len(aggregates)
-		dc.Metrics["checks"] = len(checks)
-		dc.Metrics["clients"] = len(clients)
-		dc.Metrics["events"] = len(events)
-		dc.Metrics["silenced"] = len(silenced)
-		dc.Metrics["stashes"] = len(stashes)
-		d.Data.Dc = append(d.Data.Dc, dc)
+	for _, v := range stashes {
+		setDc(v, d.datacenter.Name)
+		d.snapshot.Stashes = append(d.snapshot.Stashes, v)
 	}
+
+	d.mutex.Unlock()
+	d.wg.Done()
+}
+
+func (d *DatacenterSnapshotFetcher) fetchSilenced() {
+	silenced, err := d.datacenter.GetSilenced()
+	d.mutex.Lock()
+	if err != nil {
+		logger.Debug(err)
+		logger.Warningf("Impossible to retrieve silenced entries from the "+
+			"datacenter %s. Silencing might not be possible, please update Sensu", d.datacenter.Name)
+	}
+
+	for _, v := range silenced {
+		setDc(v, d.datacenter.Name)
+		d.snapshot.Silenced = append(d.snapshot.Stashes, v)
+	}
+
+	d.mutex.Unlock()
+	d.wg.Done()
```

Comment:
`defer d.wg.Done()` at the top of the function.

Reply:
Changed - and also made use of the mutex in `fetchEvents()`.